### ISSUE TITLE
Getting { "error": "unsupported_grant_type" }

### ISSUE THREAD

**Issue by affo** [evonove/django-oauth-toolkit#127]
Hi,
I am writing an AngularJS application that consumes mi REST Framework APIs.  
I am using DOT as oAuth2 provider.  
The problem is that the same request sent with cURL is successful, once it is sent with an XHR request I get:

    { "error": "unsupported_grant_type" }

I suspected an encoding problem, so I tried to change headers putting:

    'Content-Type': 'application/x-www-form-urlencoded;charset=utf-8'

But it doesn't work...  
For precision's sake, XHR requests are sent with Restangular.

Thank you in advance

**Comment by palazzem** [comment]
Hi (and sorry for taking so long for an answer),
the default behavior is to enforce the use of ``application/x-www-form-urlencoded`` Content-Type.
The ``OAuthLibCore`` implementation extracts the request body using ``request.POST`` dictionary, which is empty for a request that uses ``application/json`` Content-Type.

Anyhow, from your settings, you can set the ``JSONOAuthLibCore`` class that supports ``application/json``. The following example shows how to use the new backend class:

```
OAUTH2_PROVIDER = {
    # other OAUTH2 settings
    'OAUTH2_BACKEND_CLASS': 'oauth2_provider.oauth2_backends.JSONOAuthLibCore'
}
```

This feature will be released in the milestone ``0.9.0`` and is [already merged][1] in master.
Regards.

[1]: https://github.com/evonove/django-oauth-toolkit/pull/234

**Comment by sacgup** [comment]
Hi, Isn't it possible that OAUTH2_BACKEND_CLASS is used based on the content type of the request.
I actually want to support 'Content-Type': 'application/x-www-form-urlencoded;charset=utf-8' as well as 'application/json'. Is there any way i can do it ?

**Comment by theromis** [comment]
+1 I want conditional content type selector as well

**Comment by ygag-basil** [comment]
> JSONOAuthLibCore

Just create a custom oauth2_backends 

```
import json

from oauth2_provider.oauth2_backends import OAuthLibCore

class JSONOAuthLibCoreCustom(OAuthLibCore):
    """
    Extends the default OAuthLibCore to parse correctly application/json requests
    """

    def extract_body(self, request):
        try:
            body = json.loads(request.body.decode("utf-8")).items()
        except (AttributeError, ValueError):
            body = request.POST.items()

        return body
```

and 

```
OAUTH2_PROVIDER = {
    # other OAUTH2 settings
    'OAUTH2_BACKEND_CLASS': 'path.to.JSONOAuthLibCoreCustom'
}
```

**Comment by icon-fonts** [comment]
> Hi (and sorry for taking so long for an answer), the default behavior is to enforce the use of `application/x-www-form-urlencoded` Content-Type. The `OAuthLibCore` implementation extracts the request body using `request.POST` dictionary, which is empty for a request that uses `application/json` Content-Type.
> 
> Anyhow, from your settings, you can set the `JSONOAuthLibCore` class that supports `application/json`. The following example shows how to use the new backend class:
> 
> ```
> OAUTH2_PROVIDER = {
>     # other OAUTH2 settings
>     'OAUTH2_BACKEND_CLASS': 'oauth2_provider.oauth2_backends.JSONOAuthLibCore'
> }
> ```
> 
> This feature will be released in the milestone `0.9.0` and is [already merged](https://github.com/evonove/django-oauth-toolkit/pull/234) in master. Regards.





https://stackdev.space/questions/executing-a-restangular-get-request-with-information-included
